web/styles/compose.css

```diff
@@ -1292,6 +1292,12 @@ textarea.new_message_textarea {
     .typeahead-menu {
         list-style: none;
         margin: 4px 0;
+        max-height: 248px;
```

Comment:
We need to set the adjust the height based on widow height otherwise the some typeahead items will not be visible at small heights.

Reply:
I had considered setting it to something like `max-height: min(248px, 90vh);` but didn't since I don't see any practical cases where the height is under 250px... What exactly do you suggest here?

Reply:
We can use something we use for dropdown widgets here:

```

function compose_recipient_dropdown_on_show(dropdown: tippy.Instance): void {
    // Offset to display dropdown above compose.
    let top_offset = 5;
    const window_height = window.innerHeight;
    const search_box_and_padding_height = 50;
    // pixels above compose box.
    const recipient_input_top = $("#compose_select_recipient_widget_wrapper").get_offset_to_window()
        .top;
    const top_space = recipient_input_top - top_offset - search_box_and_padding_height;
    // pixels below compose starting from top of compose box.
    const bottom_space = window_height - recipient_input_top - search_box_and_padding_height;
    // Show dropdown on top / bottom based on available space.
    let placement: tippy.Placement = "top-start";
    if (bottom_space > top_space) {
        placement = "bottom-start";
        top_offset = -30;
    }
    const offset: [number, number] = [-10, top_offset];
    dropdown.setProps({placement, offset});
    const height = Math.min(
        dropdown_widget.DEFAULT_DROPDOWN_HEIGHT,
        Math.max(top_space, bottom_space),
    );
    const $popper = $(dropdown.popper);
    $popper.find(".dropdown-list-wrapper").css("max-height", height + "px");
}

```

Reply:
Hmm, being this exact might be difficult for typeaheads since they are at a variety of locations in different contexts, but i'll try something

Reply:
Since the reference element for composebox typeaheads is the textbox under which there is no space, we cannot use this strategy of keeping the typeahead either strictly above or below the reference element. 

The current tippy placement where 1 side is preferred, but the typeahead is repositioned to keep it within the viewport works better, so I've retained that and just changed `max-height: 248px` to `max-height: min(248px, 95vh)` so it fits in whatever space is available. This means it sometimes covers the refrence element but that happens anyway, so it's not a regression, and the whole typeahead fitting in the available vertical space and being scrollable is an improvement:
![image](https://github.com/zulip/zulip/assets/68962290/19694a8d-24ff-4f7e-9a9f-54c02d86aa94)

Reply:
Hope this works?

Reply:
This is not ideal but I noticed today that typeaheads don't work on mobile since the `keyup` event is not being triggered on mobile devices. I will open an issue for that, along with stream dropdown not working on mobile devices.

Reply:
Then, we can come back to improving this behaviour.

Reply:
started discussion in https://chat.zulip.org/#narrow/stream/9-issues/topic/Typeaheads.20and.20stream.20dropdown.20not.20working.20on.20mobile